### DISCUSSION TITLE
Improve Shortcut.matches_event() not matching InputEventKey events precisely

### DISCUSSION THREAD

**Discussion by Jowan-Spooner** [godotengine/godot-proposals#10758] [GUI]
As explained in https://github.com/godotengine/godot/issues/97048 `Shortcut.matches_event()` a Shortcut setup with an InputEventKey event currently ignores the `InputEventKey.pressed` property and will always return `true` for both *pressed* and *released* variants of that event. This is easy to workaround, but unintuitive. 

Intuitively I setup a shortcut like this:
<img src="https://github.com/user-attachments/assets/fe72c13a-335c-486e-8b0f-75879fe89cb4" width="300">

Then checked for the shortcut like this:
```gdscript
func _input(event:InputEvent) -> void:
	if shortcut.matches_event(event):
		do_something()
```
However this will trigger twice for any InputEventKey which is not what you want for a shortcut and feels unintuitive because you can change the `pressed` property of the InputEventKey but that has no effect unless you check it manually like this:
```gdscript
func _input(event:InputEvent) -> void:
	if shortcut.matches_event(event) and event.pressed == shortcut.events[0].pressed:
		do_something()
		accept_event()
```

Apparently this is intended (not sure what for) and standard (InputEvent.matches()) does this too. I don't know of the use cases of InputEvent.matches() but at least for shortcuts this feels strange. 

Interestingly the same is true for `InputEventKey.echo`. Also something uncommon for shortcuts and something you have to check manually. 

Maybe the most intuitive solution would be to hide `pressed`, `echo` and similar *"event state"* properties from the inspector cause editing/seeing them is probably what confused me. However I assume this is not a good solution either.

Not sure how to best solve this, but the current workflow is confusing.

**Comment by AThousandShips** [comment]
There's a simple workaround for this, allowing you to do any variation on what you want matched:
```gdscript
func my_check(list_event: InputEvent, check_event: InputEvent, exact: bool, check_pressed: bool, check_echo: bool) -> bool:
    return list_event and list_event.is_match(check_event, exact) and (not check_pressed or list_event.is_pressed() == check_event.is_pressed()) and (not check_echo or list_event.is_echo() == check_event.is_echo())

func _input(event: InputEvent) -> void:
	if shortcut.events.any(my_check.bind(event, true, true, true)):
		do_something()
```
For example, or just configure as you wish

For comparison, this is the code of the actual check:
```c++
bool Shortcut::matches_event(const Ref<InputEvent> &p_event) const {
	Ref<InputEventShortcut> ies = p_event;
	if (ies.is_valid()) {
		if (ies->get_shortcut().ptr() == this) {
			return true;
		}
	}

	for (int i = 0; i < events.size(); i++) {
		Ref<InputEvent> ie = events[i];
		bool valid = ie.is_valid() && ie->is_match(p_event);

		// Stop on first valid event - don't need to check further.
		if (valid) {
			return true;
		}
	}

	return false;
}
```

**Reply by Jowan-Spooner** [reply]
Yes, this is a good workaround 👍. 

It seems to me that you'd often (or mostly?) want to match a shortcut while also matching pressed and echo (wanting to trigger shortcuts on echo events seems to be an unlikely use-case tbh) so maybe a utility method that does a typical shortcut check (checking both pressed and echo) could be added?

Again, not really necessary, but `Shortcut.matches_event()` isn't really necessary in the first place as it can be worked around with the same method you've proposed.

**Reply by AThousandShips** [reply]
I disagree, and you'd usually check for pressed *before* checking this, that's how it is normally done in the engine itself, like so:
```gdscript
if event and event.is_pressed() and not event.is_echo():
    if shortcut_copy.matches_event(event):
        do_cut()
    elif shortcut_paste.matches_event(event):
        do_paste()
    ...
```
It's be a waste of processing to check it for *each* shortcut

**Reply by AThousandShips** [reply]
And this would be sufficient as you'd normally never mix events with and without echo or mixed pressed/released like this, having the code behave the same way both on release and on press for different events (for example, do the same action on pressing `X` and releasing `Y`), or one event with and one event without echo (for example, for only the first press of `X` but for every echo update when holding down `Y`)
